File: iosApp/iosApp/Features/Assistant/StationGrouping.swift
import Foundation
import CoreLocation

/// A presentation-level station group (finding 2): one selector row per real
/// physical station choice.
///
/// Co-located stops that share a station identity, the same folded name AND the
/// same location, collapse into a single row that carries every underlying
/// routable stop id and the union of the lines that serve them. So the selector
/// never shows the same physical station twice (e.g. the tram `A1_KIF` and
/// `A2_KIF`, both "Kifisias" at one point), while genuinely distinct stations stay
/// separate: "Kifissia" (`M1_KIF`) and "Kifisias" differ in both name and place,
/// so they never merge.
///
/// Identity is the conjunction of name AND location, never name or proximity
/// alone: two ids merge only when their folded names match and they sit at
/// essentially the same point. Selection keeps access to every member line because
/// the planner links co-located members with transfer edges, so no line becomes
/// unreachable when a group resolves to its representative stop.
struct StationGroup: Identifiable, Equatable {
    /// Stable across launches and input order: the smallest member id.
    let id: String
    let name: String
    let nameEl: String
    /// Every routable stop id in the group, sorted, so nothing is lost on merge.
    let memberIds: [String]
    /// Union of the lines serving the group, for the row's disambiguating badges.
    let lineIds: [String]
    let region: TransitRegion
    let coordinate: CLLocationCoordinate2D

    /// The id carried to the planner / stored on selection.
    var representativeId: String { id }

    static func == (a: StationGroup, b: StationGroup) -> Bool {
        a.id == b.id && a.memberIds == b.memberIds && a.lineIds == b.lineIds
    }
}

enum StationGrouping {
    /// Fold case and accents but keep the letters, so accented and unaccented
    /// spellings match for search and grouping while genuinely different spellings
    /// (Kifissia vs Kifisias) never collapse together.
    static func fold(_ s: String) -> String {
        s.folding(options: [.caseInsensitive, .diacriticInsensitive], locale: nil)
            .trimmingCharacters(in: .whitespaces)
    }

    /// ~11m coordinate bucket. Only near-identical points merge, so co-location is
    /// required in addition to a name match: proximity is evidence, not identity.
    private static func bucket(_ c: CLLocationCoordinate2D) -> String {
        String(format: "%.4f,%.4f", c.latitude, c.longitude)
    }

    /// Collapse a raw station list into presentation groups, preserving first-seen
    /// order so the picker stays stable as input is reordered.
    static func groups(from stations: [TransitStation]) -> [StationGroup] {
        var buckets: [String: [TransitStation]] = [:]
        var order: [String] = []
        for st in stations {
            let key = fold(st.name.isEmpty ? st.nameEl : st.name) + "|" + bucket(st.coordinate)
            if buckets[key] == nil { order.append(key) }
            buckets[key, default: []].append(st)
        }
        return order.map { key in
            let members = buckets[key]!.sorted { $0.id < $1.id }
            let rep = members.first!
            var lineIds: [String] = []
            for m in members { for l in m.lineIds where !lineIds.contains(l) { lineIds.append(l) } }
            return StationGroup(
                id: rep.id, name: rep.name, nameEl: rep.nameEl,
                memberIds: members.map { $0.id }, lineIds: lineIds,
                region: rep.region, coordinate: rep.coordinate)
        }
    }

    /// Whether a group matches a folded query on any of its member names (English
    /// or Greek), so accented and unaccented input both hit.
    static func matches(_ group: StationGroup, query folded: String) -> Bool {
        if folded.isEmpty { return true }
        return fold(group.name).contains(folded) || fold(group.nameEl).contains(folded)
    }
}

File: core/domain/src/commonMain/kotlin/com/syrmos/core/domain/station/StationGrouping.kt
package com.syrmos.core.domain.station

import com.syrmos.core.model.transit.Station
import kotlin.math.round

/**
 * A presentation-level station group (finding 2, Android/shared peer of the iOS
 * `StationGrouping`): one selector row per real physical station.
 *
 * Co-located stops that share a station identity, the same folded name AND the
 * same location, collapse into one group that carries every underlying stop id and
 * the union of the lines serving them. So the selector never lists the same
 * physical station twice (the tram `A1_KIF` / `A2_KIF`, both "Kifisias" at one
 * point), while genuinely distinct stations stay separate: "Kifissia" and
 * "Kifisias" differ in both name and place, so they never merge. Identity is the
 * conjunction of name AND location, never one alone.
 */
data class StationGroup(
    /** Stable across launches and input order: the smallest member id. */
    val id: String,
    val name: String,
    val nameEl: String,
    val nameSq: String?,
    /** Every routable stop id in the group, sorted, so nothing is lost on merge. */
    val memberIds: List<String>,
    /** Union of the lines serving the group, for the row's disambiguating badges. */
    val lineIds: List<String>,
) {
    /** The id carried to the planner on selection. */
    val representativeId: String get() = id
}

object StationGrouping {

    // Fold case and accents but keep the letters, so accented and unaccented
    // spellings match while genuinely different spellings never collapse together.
    private val FOLD: Map<Char, Char> = mapOf(
        'ά' to 'α', 'έ' to 'ε', 'ή' to 'η', 'ί' to 'ι', 'ό' to 'ο', 'ύ' to 'υ', 'ώ' to 'ω',
        'ϊ' to 'ι', 'ϋ' to 'υ', 'ΐ' to 'ι', 'ΰ' to 'υ',
        'á' to 'a', 'é' to 'e', 'í' to 'i', 'ó' to 'o', 'ú' to 'u', 'à' to 'a', 'è' to 'e',
        'ç' to 'c', 'ë' to 'e',
    )

    fun fold(s: String): String {
        val lower = s.trim().lowercase()
        val sb = StringBuilder(lower.length)
        for (c in lower) sb.append(FOLD[c] ?: c)
        return sb.toString()
    }

    // ~11m coordinate bucket. Only near-identical points merge, so co-location is
    // required in addition to a name match: proximity is evidence, not identity.
    private fun bucket(lat: Double, lon: Double): String =
        "${round(lat * 10000.0) / 10000.0},${round(lon * 10000.0) / 10000.0}"

    /** Collapse a raw station list into presentation groups, first-seen order. */
    fun groups(stations: List<Station>): List<StationGroup> {
        val buckets = LinkedHashMap<String, MutableList<Station>>()
        for (st in stations) {
            val key = fold(st.name.ifEmpty { st.nameEl }) + "|" + bucket(st.latitude, st.longitude)
            buckets.getOrPut(key) { mutableListOf() }.add(st)
        }
        return buckets.values.map { raw ->
            val members = raw.sortedBy { it.id }
            val rep = members.first()
            val lineIds = LinkedHashSet<String>().apply { members.forEach { addAll(it.lineIds) } }.toList()
            StationGroup(rep.id, rep.name, rep.nameEl, rep.nameSq, members.map { it.id }, lineIds)
        }
    }

    /** Whether a group matches a folded query on any of its member names. */
    fun matches(group: StationGroup, foldedQuery: String): Boolean {
        if (foldedQuery.isEmpty()) return true
        return fold(group.name).contains(foldedQuery) ||
            fold(group.nameEl).contains(foldedQuery) ||
            (group.nameSq?.let { fold(it).contains(foldedQuery) } ?: false)
    }
}
